pkg/controller/operators/adoption_controller.go

```diff
@@ -141,51 +138,31 @@ func (r *AdoptionReconciler) ReconcileSubscription(req ctrl.Request) (reconcile.
 		return reconcile.Result{}, nil
 	}
 
-	// Ensure the subscription is adopted
+	// Adopt the Subscription
 	var errs []error
-	out := in.DeepCopy()
-	adopted, err := operator.AdoptComponent(out)
-	if err != nil {
+	if err := r.adopt(ctx, operator, in); err != nil {
+		log.Error(err, "Error adopting Subscription")
 		errs = append(errs, err)
 	}
-	if adopted {
-		if err := r.Patch(ctx, out, client.MergeFrom(in)); err != nil {
-			log.Error(err, "Error adopting Subscription")
-			errs = append(errs, err)
-		}
-	}
 
-	// Find the Subscription's CSVs and apply the component label if necessary
-	adoptCSV := func(name string) error {
+	// Adopt the Subscription's installed CSV
+	if name := in.Status.InstalledCSV; name != "" {
 		csv := &operatorsv1alpha1.ClusterServiceVersion{}
-		if err := r.Get(ctx, types.NamespacedName{Namespace: in.GetNamespace(), Name: name}, csv); err != nil {
-			if apierrors.IsNotFound(err) {
-				err = nil
-			}
-
-			return err
-		}
-		log.Info("found CSV")
-
-		candidate := csv.DeepCopy()
-		adopted, err := operator.AdoptComponent(csv)
-		if err != nil {
-			return err
-		}
-
-		if adopted {
-			// Only update the CSV if freshly adopted
-			if err := r.Patch(ctx, csv, client.MergeFrom(candidate)); err != nil {
-				return err
-			}
+		csv.SetNamespace(in.GetNamespace())
+		csv.SetName(name)
+		if err := r.adopt(ctx, operator, csv); err != nil {
+			log.Error(err, "Error adopting installed CSV")
+			errs = append(errs, err)
 		}
-
-		return nil
 	}
 
-	if name := in.Status.InstalledCSV; name != "" {
-		if err := adoptCSV(name); err != nil {
-			log.Error(err, "Error adopting installed CSV")
+	// Adopt the Subscription's latest InstallPlan
+	if ref := in.Status.InstallPlanRef; ref != nil {
+		ip := &operatorsv1alpha1.InstallPlan{}
+		ip.SetNamespace(ref.Namespace)
+		ip.SetName(ref.Name)
+		if err := r.adopt(ctx, operator, ip); err != nil {
```

Comment:
Am I missing where the installplan would be "unadopted" when it is no longer the current one?

Reply:
Nope, we don't disown InstallPlans. Is there a reason to do so now that we have a cap on the number that can exist at a given time?